tests/python/tests/test_xml_parser.py

```diff
@@ -1156,3 +1170,187 @@ def warningcb(type, msg):
+        for (pkg, name) in zip(package_iterator, package_names):
+            self.assertEqual(pkg.name, name)
+            # TODO: might be good to test that iter.is_finished() returns false while still processing packages
+
+        self.assertListEqual(warnings, [])
+        self.assertIsNone(next(package_iterator))
+        self.assertTrue(package_iterator.is_finished())
+
+    def test_xml_parser_pkg_iterator_repo02_newpkgcb_as_filter(self):
+        def newpkgcb(pkgId, name, arch):
+            if name in {"fake_bash"}:
+                return cr.Package()
+
+        package_iterator = cr.PackageIterator(
+            primary_path=REPO_02_PRIXML, filelists_path=REPO_02_FILXML, other_path=REPO_02_OTHXML,
+            newpkgcb=newpkgcb,
+        )
+
+        packages = list(package_iterator)
+
+        self.assertEqual(len(packages), 1)
+        self.assertEqual(packages[0].name, "fake_bash")
+        self.assertIsNone(next(package_iterator))
+        self.assertTrue(package_iterator.is_finished())
+
+    def test_xml_parser_pkg_iterator_warnings(self):
+        warnings = []
+        def warningcb(warn_type, msg):
+            warnings.append((warn_type, msg))
+
+        package_iterator = cr.PackageIterator(
+            primary_path=PRIMARY_MULTI_WARN_00_PATH, filelists_path=FILELISTS_MULTI_WARN_00_PATH, other_path=OTHER_MULTI_WARN_00_PATH,
+            warningcb=warningcb,
+        )
+
+        package_names = ['fake_bash', 'super_kernel']
+        for (pkg, name) in zip(package_iterator, package_names):
+            self.assertEqual(pkg.name, name)
+            # TODO: might be good to test that iter.is_finished() returns false while still processing packages
+
+        self.assertListEqual(warnings, [])
+        self.assertIsNone(next(package_iterator))
+        self.assertTrue(package_iterator.is_finished())
+
+        self.assertEqual(warnings,
+            [(0, 'Unknown element "fooelement"'),
+             (1, 'Missing attribute "type" of a package element'),
+             (0, 'Unknown element "foo"'),
+             (3, 'Conversion of "foobar" to integer failed'),
+             (0, 'Unknown element "bar"'),
+             (1, 'Missing attribute "arch" of a package element'),
+             (2, 'Unknown file type "xxx"'),
+             (0, 'Unknown element "bar"'),
+             (1, 'Missing attribute "name" of a package element'),
+             (0, 'Unknown element "bar"'),
+             (3, 'Conversion of "xxx" to integer failed')])
+
+
+    def test_xml_parser_package_iterator_error(self):
+
+        package_iterator = cr.PackageIterator(
+            primary_path=PRIMARY_ERROR_00_PATH, filelists_path=FILELISTS_ERROR_00_PATH, other_path=OTHER_ERROR_00_PATH,
+        )
+
+        with self.assertRaises(cr.CreaterepoCError) as ctx:
+            packages = list(package_iterator)
+
+        self.assertIsNone(next(package_iterator))
+        self.assertTrue(package_iterator.is_finished())
```

Comment:
The `package_iterator` already raised an error, calling `next` on it again is not a good idea. 
But maybe we could fix it so that is doesn't crash? Not sure in what state an object should be after it raises an exception.
```suggestion
```